lib/cuda_gauge_field.cu

```diff
@@ -525,33 +535,47 @@ namespace quda {
 	qudaMemcpy(gauge, buffer, bytes, cudaMemcpyHostToDevice);
 	pool_pinned_free(buffer);
       } else { // else on the GPU
-	void *buffer = create_gauge_buffer(src.Bytes(), src.Order(), src.Geometry());
-	size_t ghost_bytes[8];
-	int srcNinternal = src.Reconstruct() != QUDA_RECONSTRUCT_NO ? src.Reconstruct() : 2*nColor*nColor;
-	for (int d=0; d<geometry; d++) ghost_bytes[d] = nFace * surface[d%4] * srcNinternal * src.Precision();
-	void **ghost_buffer = (nFace > 0) ? create_ghost_buffer(ghost_bytes, src.Order(), geometry) : nullptr;
 
-	if (src.Order() == QUDA_QDP_GAUGE_ORDER) {
-	  for (int d=0; d<geometry; d++) {
-	    qudaMemcpy(((void**)buffer)[d], ((void**)src.Gauge_p())[d], src.Bytes()/geometry, cudaMemcpyHostToDevice);
+	if (src.Order() == QUDA_MILC_SITE_GAUGE_ORDER) {
+	  // special case where we use zero-copy memory to read/write directly from MILC's data
+	  void *src_d;
+	  cudaHostGetDevicePointer(&src_d, const_cast<void*>(src.Gauge_p()), 0);
```

Comment:
Until I had enough coffee just let me ask: What if `cudaHostGetDevicePointer` goes wrong?

I might need to look at the code more closely but I thought that with UVA we can get by without the `cudaHostGetDevicePointer` if we used cudaHostRegister / cudaHostMalloc.

Reply:
I've added an error on the return of this function to check that this mapping is valid.  You are correct with UVA the API call isn't needed, but no harm in having this there since it would then work on non-UVA systems (Windows anyone 😉  ?).  Also, since there's now an explicit error check on this, we'd catch now with a sensible error message if a regular pointer was passed in as opposed to a CUDA registered one (asopposedd to a random segmentation fault).

Reply:
I added the check also in another place. Using `cudaHostGetDevicePointer` to make sure the memory has been registered is a reasonable thing.